skfp/utils/splitters.py

```diff
@@ -0,0 +1,239 @@
+from collections import defaultdict
+from collections.abc import Sequence
+from typing import Union
+
+import numpy as np
+from rdkit.Chem import Mol
+from rdkit.Chem.Scaffolds import MurckoScaffold
+
+from skfp.utils.validators import ensure_mols
+
+
+def scaffold_train_test_split(
+    data: Sequence[Union[str, Mol]],
+    train_size: float = 0.8,
+    test_size: float = 0.2,
+    return_indices: bool = False,
+    include_chirality: bool = False,
+) -> Union[
+    tuple[list[str], list[str]],
+    tuple[list[Mol], list[Mol]],
+    tuple[list[int], list[int]],
+]:
+    """
+    Split a list of SMILES or RDKit 'Mol' objects into train and test subsets using Bemis-Murcko [1]_ scaffolds.
+
+    This function ensures that similar molecules (sharing the same scaffold) are in the same split.
+    This method, suggested in the MoleculeNet [2]_ paper, helps to test the model's ability to
+    generalize to entirely new scaffolds.
+
+    The split is fully deterministic, with the smallest scaffold sets assigned to the test
+    subset and the largest to the training subset.
```

Comment:
Maybe "and the rest to the training subset"? Because we can have kind of "average-sized" scaffold sets, which also get put into the training subset